File: CSharp/Client/Wharehouse.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace ThingModel
{
    public class Wharehouse
    {
        private readonly ConcurrentDictionary<string, ThingType> _thingTypes = new ConcurrentDictionary<string, ThingType>();

        private readonly ConcurrentDictionary<string, Thing> _things = new ConcurrentDictionary<string, Thing>();

        private readonly HashSet<IThingObserver> _thingObservers = new HashSet<IThingObserver>();
 
        private readonly HashSet<IThingTypeObserver> _thingTypeObservers = new HashSet<IThingTypeObserver>();
 
        public void RegisterType(ThingType type, bool force = true)
        {
            if (type == null)
            {
                throw new Exception("The thing type information is missing.");
            }

            if (force || !_thingTypes.ContainsKey(type.Name))
            {
                _thingTypes[type.Name] = type;

                NotifyThingTypeDefine(type);
            }
            
        }

        public void RegisterThing(Thing thing, bool alsoRegisterConnections = true, bool alsoRegisterTypes = false)
        {
            if (thing == null)
            {
                throw new Exception("Null are not allowed in the wharehouse.");
            }

           var creation = !_things.ContainsKey(thing.ID);
            _things[thing.ID] = thing;

            if (alsoRegisterTypes)
            {
                RegisterType(thing.Type, false);
            }


            if (alsoRegisterConnections)
            {
                var set = new HashSet<Thing>();
                foreach (var connectedThing in thing.ConnectedThings)
                {
                    RecursiveRegisterThing(connectedThing, alsoRegisterTypes, set);
                }
            }

            if (creation)
            {
                NotifyThingCreation(thing);
            }
            else
            {
                NotifyThingUpdate(thing);
            }
            
        }

        private void RecursiveRegisterThing(Thing thing, bool alsoRegisterTypes, HashSet<Thing> alreadyRegisteredThings)
        {
            
            // Avoid infinite recursions
            if (alreadyRegisteredThings.Contains(thing))
            {
                return;
            }

            alreadyRegisteredThings.Add(thing);
            
            RegisterThing(thing, false, alsoRegisterTypes);
            
            foreach (var connectedThing in thing.ConnectedThings)
            {
                RecursiveRegisterThing(connectedThing, alsoRegisterTypes, alreadyRegisteredThings);
            }
            
        }

        public void RegisterCollection(IEnumerable<Thing> collection, bool alsoRegisterTypes = false)
        {
            var set = new HashSet<Thing>();
            foreach (var thing in collection)
            {
                RecursiveRegisterThing(thing, alsoRegisterTypes, set);
            }
        }

        public void RemoveThing(Thing thing)
        {
            if (thing == null)
            {
                return;
            }

            // Remove all the connections
            foreach (var t in _things)
            {
                if (t.Value.IsConnectedTo(thing))
                {
                    t.Value.Disconnect(thing);
                    NotifyThingUpdate(t.Value);
                }
            }

            // Remove the thing
            Thing uselessThing;
            if (_things.TryRemove(thing.ID, out uselessThing))
            {
                NotifyThingDeleted(thing);
            }
        }

        public void RegisterObserver(IThingObserver observer)
        {
            _thingObservers.Add(observer);
        }

        public void RegisterObserver(IThingTypeObserver observer)
        {
            _thingTypeObservers.Add(observer);
        }

        public void Unregister(IThingObserver observer)
        {
            _thingObservers.Remove(observer);
        }

        public void Unregister(IThingTypeObserver observer)
        {
            _thingTypeObservers.Remove(observer);
        }

        public void NotifyThingTypeDefine(ThingType type)
        {
            foreach (var observer in _thingTypeObservers)
            {
                observer.Define(type);
            }
        }

        public void NotifyThingUpdate(Thing thing)
        {
            foreach (var observer in _thingObservers)
            {
                observer.Updated(thing);
            }
        }

        public void NotifyThingCreation(Thing thing)
        {
            foreach (var observer in _thingObservers)
            {
                observer.New(thing);
            }
        }

        public void NotifyThingDeleted(Thing thing)
        {
            foreach (var observer in _thingObservers)
            {
                observer.Deleted(thing);
            }
        }

        public Thing GetThing(string key)
        {
            Thing value;
            _things.TryGetValue(key, out value);
            return value;
        }

        public ThingType GetThingType(string key)
        {
            ThingType value;
            _thingTypes.TryGetValue(key, out value);
            return value;
        }

        public IList<Thing> Things { get { return new List<Thing>(_things.Values); }
        }
    }
}

File: Java/src/main/java/org/thingmodel/Wharehouse.java
package org.thingmodel;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class Wharehouse {

	private HashMap<String, ThingType> _thingTypes = new HashMap<>();
	private HashMap<String, Thing> _things = new HashMap<>();
	private HashSet<IWharehouseObserver> _observers = new HashSet<>();
	
	private final Object _lockThingTypes = new Object();
	private final Object _lockThings = new Object();
	
	
	public void RegisterType(ThingType type) {
		RegisterType(type, true);
	}
	
	public void RegisterType(ThingType type, boolean force) {
		if (type == null) {
			throw new RuntimeException("The ThingType object should not be null");
		}
	
		String name = type.getName();
		boolean register = force;
		
		if (!register) {
			synchronized (_lockThingTypes) {
				register = !_thingTypes.containsKey(name);
			}
		}
		
		if (register) {
			
			synchronized (_lockThingTypes) {
				_thingTypes.put(name, type);
			}
			
			NotifyThingTypeDefine(type);
		}
	}

	public void RegisterThing(Thing thing) {
		RegisterThing(thing, true, false);
	}
	
	public void RegisterThing(Thing thing, boolean alsoRegisterConnections, boolean alsoRegisterTypes) {
		if (thing == null) {
			throw new RuntimeException("Null things are not allowed in the warehouse.");
		}
		
		String id = thing.getId();
		boolean creation;
	
		synchronized (_lockThings) {
			creation = !_things.containsKey(id);
			_things.put(id, thing);
		}
		
		if (alsoRegisterTypes && thing.getType() != null) {
			RegisterType(thing.getType(), false);
		}
		
		if (alsoRegisterConnections) {
			HashSet<String> alreadyVisitedThings = new HashSet<>();
			
			for (Thing connectedThing : thing.getConnectedThings()) {
				RecursiveRegisterThing(connectedThing, alsoRegisterTypes, alreadyVisitedThings);
			}
		}
		
		if (creation) {
			NotifyThingCreation(thing);
		} else {
			NotifyThingUpdate(thing);
		}
		
	}
	
	
	private void RecursiveRegisterThing(Thing thing, boolean alsoRegisterTypes,
			HashSet<String> alreadyVisitedThings) {
	
		String id = thing.getId();
		
		// Avoid infinite recursions
		if (alreadyVisitedThings.contains(id)) {
			return;
		}
		
		alreadyVisitedThings.add(id);
		
		RegisterThing(thing, false, alsoRegisterTypes);
		
		for(Thing connectedThing : thing.getConnectedThings()) {
			RecursiveRegisterThing(connectedThing, alsoRegisterTypes, alreadyVisitedThings);
		}
		
	}
	
	public void RegisterCollection(Collection<Thing> collection) {
		RegisterCollection(collection, false);
	}
	
	public void RegisterCollection(Collection<Thing> collection, boolean alsoRegisterTypes) {
		HashSet<String> alreadyVisitedThings = new HashSet<>();
		for (Thing thing : collection) {
			this.RecursiveRegisterThing(thing, alsoRegisterTypes, alreadyVisitedThings);
		}
	}
	
	public void RemoveThing(Thing thing) {
		if (thing == null) {
			return;
		}
		
		ArrayList<Thing> thingsToDisconnect = new ArrayList<>();
	
		// Remove all connections
		synchronized (_lockThings) {
			for(Thing t : _things.values()) {
				if (t.IsConnectedTo(thing)) {
                    thingsToDisconnect.add(t);
				}
			}
		}

		boolean removed;
		String id = thing.getId();
		
		synchronized (_lockThings) {
			removed = _things.containsKey(id);
			if (removed) {
				_things.remove(id);
			}
		}
		
		if (removed) {
			NotifyThingDeleted(thing);
		}
		
		for (Thing t : thingsToDisconnect) {
			t.Disconnect(thing);
			NotifyThingUpdate(t);
		}
	}
	
	public void RegisterObserver(IWharehouseObserver observer) {
		_observers.add(observer);
	}
	
	public void UnregisterObserver(IWharehouseObserver observer) {
		_observers.remove(observer);
	}
	
	public void NotifyThingTypeDefine(ThingType type) {
		for (IWharehouseObserver observer : _observers) {
			observer.Define(type);
		}
	}
	
	public void NotifyThingCreation(Thing thing) {
		for (IWharehouseObserver observer : _observers) {
			observer.New(thing);
		}
	}
	
	public void NotifyThingUpdate(Thing thing) {
		for (IWharehouseObserver observer : _observers) {
			observer.Updated(thing);
		}
	}

	public void NotifyThingDeleted(Thing thing) {
		for (IWharehouseObserver observer : _observers) {
			observer.Deleted(thing);
		}
	}
	
	public Thing getThing(String id) {
		synchronized (_lockThings) {
			return _things.get(id);
		}
	}
	
	public ThingType getThingType(String name) {
		synchronized (_lockThingTypes) {
			return _thingTypes.get(name);
		}
	}
	
	public List<Thing> getThings() {
		synchronized (_lockThings) {
			return new ArrayList<>(_things.values());
		}
	}
}
